reviews/tests.py

```diff
@@ -267,3 +300,168 @@ def test_auto_review_task(self):
 
         reasons = auto_review_task(self.study, s1_t1)
         self.assertEqual(len(reasons), 1)
+
+
+class BaseViewTestCase(BaseReviewTestCase):
+
+    # This testcase supports only class-based views
+    view_class = None
+
+    # for example:
+    # "/proposals/update/1/"
+    # NOT a full URL including protocal and domain
+    view_path = None
+
+    allowed_users = []
+    disallowed_users = [AnonymousUser]
+    enforce_csrf = True
+
+    def setUp(self):
+        super().setUp()
+        self.client = Client()
+        self.view = self.view_class.as_view()
+        self.factory = RequestFactory()
+        self.review = start_review(self.proposal)
+
+    def check_access(self, user):
+        request = self.factory.get(
+            self.get_view_path(),
+        )
+        request.user = user
+        response = self.view(request, pk=self.review.pk)
+        return response.status_code == 200
+
+    def post(self, update_dict={}):
+        """Generic function to test form submission"""
+        post_data = {}
+        post_data.update(update_dict)
+        if self.enforce_csrf:
+            csrf_token = self.fetch_csrf_token(
+                user=self.secretary,
+            )
+            post_data["csrfmiddlewaretoken"] = csrf_token
+        response = self.client.post(
+            self.get_view_path(),
+            data=post_data,
+        )
+        return response
+
+    def fetch_csrf_token(self, user=None):
+        if user:
+            self.client.force_login(user)
+        page = self.client.get(
+            self.get_view_path(),
+        )
+        return page.context["csrf_token"]
+
+    def get_view_path(self):
+        return self.view_path
+
+
+class ReviewCloseTestCase(BaseViewTestCase):
+
+    view_class = ReviewCloseView
+
+    def get_view_path(self):
+        pk = self.review.pk
+        return f"/reviews/close/{pk}/"
```

Comment:
Because I want the test to fail if the literal URL changes.

Most likely scenario is that such a URL change would be intentional. But it's a good canary for unintentional URL changes. Those would slip under the radar if we used `reverse()` here. Probably not so important for this view specifically, but good practice for tests in general IMO.

Added a comment to reflect this.

Reply:
I can see where you're coming from, but I still think it's strange;

> Because I want the test to fail if the literal URL changes.

If you want to test something, then write a test for that. ;)
The thing is, this method is not a test itself. Thus, seemingly unrelated tests will fail if the url changes, which would make me mildly annoyed even if it's documented in the testcase itself.

If your objective is to make sure the URL doesn't change with a test, I would prefer to make it an actual dedicated testcase. (And use reverse everywhere else). I think it's better practice to keep your tests as focused as possible and avoid other problems from affecting a specific testcase. That way it's more obvious why we have a failing test, thus accomplishing the goal of testing more effectively. 

Also, if we were to have a specific ``TestCase`` class for URLs, it would give us an oppertunity to add testcases for URLs that are used outside the app (or in emails). Because you're right, we should be checking that ;)

Reply:
On second thought, it actually does make sense to test the `reverse()` functionality for this view as well. That would be a canary for an unintentional view name (i.e. `reviews:close`) change. But I'm a little conflicted on whether that belongs in this test or in the test for the views that actually call `reverse("reviews:close")`. Will think about it some more

Reply:
Added a separate routing test to cover both cases